File: src/main/java/io/stouder/adonis/edge/file/EdgeTemplateHighlighter.java
package io.stouder.adonis.edge.file;

import com.intellij.lang.Language;
import com.intellij.lang.javascript.JavaScriptFileType;
import com.intellij.openapi.editor.colors.EditorColorsScheme;
import com.intellij.openapi.editor.ex.util.LayerDescriptor;
import com.intellij.openapi.editor.ex.util.LayeredLexerEditorHighlighter;
import com.intellij.openapi.fileTypes.*;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.templateLanguages.TemplateDataLanguageMappings;
import io.stouder.adonis.edge.EdgeLanguage;
import io.stouder.adonis.edge.highlight.EdgeHighlighter;
import io.stouder.adonis.edge.parsing.EdgeTokenTypes;

public class EdgeTemplateHighlighter extends LayeredLexerEditorHighlighter {
    public EdgeTemplateHighlighter(Project project, VirtualFile virtualFile, EditorColorsScheme scheme) {
        super(new EdgeHighlighter(), scheme);

        LanguageFileType languageFileType = null;
        if(project == null || virtualFile == null) {
            languageFileType = FileTypes.PLAIN_TEXT;
        } else {
            Language language = TemplateDataLanguageMappings.getInstance(project).getMapping(virtualFile);
            if(language != null) languageFileType = language.getAssociatedFileType();
            if(languageFileType == null) languageFileType = EdgeLanguage.getDefaultTemplateLang();
        }

        SyntaxHighlighter outerHighlighter = SyntaxHighlighterFactory.getSyntaxHighlighter(languageFileType, project, virtualFile);
        SyntaxHighlighter javascriptHighlighter = SyntaxHighlighterFactory.getSyntaxHighlighter(JavaScriptFileType.INSTANCE, project, virtualFile);

        assert outerHighlighter != null;
        assert javascriptHighlighter != null;

        this.registerLayer(EdgeTokenTypes.CONTENT, new LayerDescriptor(outerHighlighter, ""));
        this.registerLayer(EdgeTokenTypes.MUSTACHE_CONTENT, new LayerDescriptor(javascriptHighlighter, ""));
        this.registerLayer(EdgeTokenTypes.TAG_CONTENT, new LayerDescriptor(javascriptHighlighter, ""));
    }
}

File: src/main/java/io/stouder/adonis/edge/file/EdgeTemplateHighlighter.kt
package io.stouder.adonis.edge.file

import com.intellij.lang.Language
import com.intellij.lang.javascript.JavaScriptFileType
import com.intellij.openapi.editor.colors.EditorColorsScheme
import com.intellij.openapi.editor.ex.util.LayerDescriptor
import com.intellij.openapi.editor.ex.util.LayeredLexerEditorHighlighter
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.fileTypes.FileTypeRegistry
import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.templateLanguages.TemplateDataLanguageMappings
import io.stouder.adonis.edge.EdgeLanguage
import io.stouder.adonis.edge.highlight.EdgeHighlighter
import io.stouder.adonis.edge.parsing.EdgeTokenTypes

class EdgeTemplateHighlighter(project: Project?, virtualFile: VirtualFile?, scheme: EditorColorsScheme) :
    LayeredLexerEditorHighlighter(EdgeHighlighter(), scheme) {

    init {
        var languageFileType: FileType?
        if (project == null || virtualFile == null) {
            languageFileType = FileTypeRegistry.getInstance().getFileTypeByFileName(virtualFile?.name ?: "")
        } else {
            val language = TemplateDataLanguageMappings.getInstance(project).getMapping(virtualFile)
            languageFileType = language?.associatedFileType
            if (languageFileType == null) {
                languageFileType = EdgeLanguage.getDefaultTemplateLang()
            }
        }

        val outerHighlighter = SyntaxHighlighterFactory.getSyntaxHighlighter(languageFileType, project, virtualFile)
        val javascriptHighlighter = SyntaxHighlighterFactory.getSyntaxHighlighter(JavaScriptFileType.INSTANCE, project, virtualFile)

        checkNotNull(outerHighlighter) { "Failed to get syntax highlighter for languageFileType: $languageFileType" }
        checkNotNull(javascriptHighlighter) { "Failed to get syntax highlighter for JavaScriptFileType" }

        registerLayer(EdgeTokenTypes.CONTENT, LayerDescriptor(outerHighlighter, ""))
        registerLayer(EdgeTokenTypes.MUSTACHE_CONTENT, LayerDescriptor(javascriptHighlighter, ""))
        registerLayer(EdgeTokenTypes.TAG_CONTENT, LayerDescriptor(javascriptHighlighter, ""))
    }
}
